RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc

```diff
@@ -43,31 +43,38 @@ PatternRecognitionbyCA::PatternRecognitionbyCA(const edm::ParameterSet &conf, co
   eidSession_ = tensorflow::createSession(trackstersCache->eidGraphDef);
 }
 
-PatternRecognitionbyCA::~PatternRecognitionbyCA(){};
+template <typename TILES>
+PatternRecognitionbyCA<TILES>::~PatternRecognitionbyCA(){};
 
-void PatternRecognitionbyCA::makeTracksters(const PatternRecognitionAlgoBase::Inputs &input,
-                                            std::vector<Trackster> &result,
-                                            std::unordered_map<int, std::vector<int>> &seedToTracksterAssociation) {
+template <typename TILES>
+void PatternRecognitionbyCA<TILES>::makeTracksters(
+    const typename PatternRecognitionAlgoBaseT<TILES>::Inputs &input,
+    std::vector<Trackster> &result,
+    std::unordered_map<int, std::vector<int>> &seedToTracksterAssociation) {
   // Protect from events with no seeding regions
   if (input.regions.empty())
     return;
 
   rhtools_.getEventSetup(input.es);
 
-  theGraph_->setVerbosity(algo_verbosity_);
+  theGraph_->setVerbosity(PatternRecognitionAlgoBaseT<TILES>::algo_verbosity_);
   theGraph_->clear();
-  if (algo_verbosity_ > None) {
+  if (PatternRecognitionAlgoBaseT<TILES>::algo_verbosity_ > PatternRecognitionAlgoBaseT<TILES>::None) {
     LogDebug("HGCPatternRecoByCA") << "Making Tracksters with CA" << std::endl;
   }
 
+  int type = input.tiles[0].getTypeT();
```

Comment:
from a design point of view, it would be better to associate the constants with the tile type in a traits class
then the nEtaBins and nPhiBins values would be known at compile time